py/dml/codegen.py

```diff
@@ -394,6 +395,462 @@ def memoization_common_prelude(name, site, outp, throws, make_ref):
+    def __init__(self):
+        self._dict = {}
+        self._list = []
+
+    def get(self, index, realtyped=False, createnew=False):
+        index = tuple(index if realtyped
+                      else (safe_realtype(typ) for typ in index))
+        classes = tuple(type(elem) for elem in index)
+        try:
+            infos = self._dict[classes]
+        except KeyError:
+            if createnew:
+                info = TypeSequenceInfo(index, len(self._list))
+                self._dict[classes] = [info]
+                self._list.append(info)
+                return info
+            else:
+                return None
+        for info in infos:
+            if (len(info.types) == len(index)
+                and all(typK.cmp(typI) == 0
+                        for (typK, typI) in zip(info.types, index))):
+                return info
+        if createnew:
+            info = TypeSequenceInfo(index, len(self._list))
+            self._list.append(info)
+            return info
+        else:
+            return None
+
+    def items(self):
+        return self._list
+
+class AfterInfo(metaclass=ABCMeta):
+    def __init__(self, key, dimsizes):
+        self.key = key
+        self.dimsizes = dimsizes
+
+    @abstractproperty
+    def string_key(self): pass
+
+    @abstractproperty
+    def cident_prefix(self): pass
+
+    @abstractproperty
+    def args_type(self): pass
+
+    @abstractproperty
+    def types_to_declare(self): pass
+
+    @property
+    def cident_callback(self):
+        return self.cident_prefix + 'callback'
+
+    @property
+    def dimensions(self):
+        return len(self.dimsizes)
+
+class AfterDelayInfo(AfterInfo):
+    def __init__(self, key, dimsizes, uniq):
+        self.uniq = uniq
+        super().__init__(key, dimsizes)
+
+    @abstractmethod
+    def generate_callback_call(self, indices_lit, args_lit): pass
+
+    @property
+    def cident_prefix(self):
+        return f'_simple_event_{self.uniq}_'
+
+    @property
+    def cident_get_value(self):
+        if self.dimensions or self.args_type is not None:
+            return self.cident_prefix + 'get_value'
+        else:
+            return '_simple_event_only_domains_get_value'
+
+    @property
+    def cident_set_value(self):
+        if self.dimensions or self.args_type is not None:
+            return self.cident_prefix + 'set_value'
+        else:
+            return '_simple_event_only_domains_set_value'
+
+
+class AfterOnHookInfo(AfterInfo):
+    def __init__(self, dimsizes, parent, typeseq_info, prim_key,
+                 param_to_msg_comp, inp, has_serialized_args):
+        self.typeseq_info = typeseq_info
+        self.parent = parent
+        # TODO drop?
+        self.inp = inp
+        self.has_serialized_args = has_serialized_args
+        self.prim_key = prim_key
+        self.param_to_msg_comp = param_to_msg_comp
+        self.uniq = len(dml.globals.after_on_hook_infos)
+        dml.globals.after_on_hook_infos.append(self)
```

Comment:
I tend to prefer side-effect free constructors; create a separate global function for instantiation+registration if needed, to allow separation of "real" instances from unit-test instances. Not a strong opinion, will not insist on it.

Reply:
Getting a uniq is so critical to the workings of `AfterOnHookInfo` and everything interacting with it that it doesn't make sense to me to separate it.